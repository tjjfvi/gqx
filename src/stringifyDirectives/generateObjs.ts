
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
