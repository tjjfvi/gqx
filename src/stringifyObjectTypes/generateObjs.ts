
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
