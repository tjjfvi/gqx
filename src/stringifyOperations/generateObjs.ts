
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
