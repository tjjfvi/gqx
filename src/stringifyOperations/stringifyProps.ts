
import { Obj } from ".";
import stringifyId from "./stringifyId";
import indent from "../indent";

const str = (x: string) => JSON.stringify(x);

export default (objs: Obj[]) =>
  objs.map(obj => obj.props.map(prop => `export class ${stringifyId(prop.id)} {
  private static _: any;
  static typeProp = ${str(obj.prop)};
  static type = ${str(obj.type)};
  static prop = ${str(prop.id.prop)};
  static inputTypes = {${prop.args.length ? "\n" + prop.args.map(n =>
    indent(indent`${n.name.value}: ${str(n.type.loc.source.body.slice(n.type.loc.start, n.type.loc.end))},\n`)
  ).join("") + "  " : ""}};
}\n\n`).join("")).join("");
