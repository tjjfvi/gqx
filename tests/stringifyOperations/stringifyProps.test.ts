import generateObjs from "../../src/stringifyOperations/generateObjs"
import stringifyProps from "../../src/stringifyOperations/stringifyProps";

test("", () => {
  let [objs] = generateObjs(utils.getOperations(["extensive"]));
  let code = stringifyProps(objs);
  expect(code).toParse().toMatchSnapshot();
})
