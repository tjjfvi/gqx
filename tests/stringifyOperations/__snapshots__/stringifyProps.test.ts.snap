// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"class Query$getAuthor {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"getAuthor\\";
  static inputTypes = {
    id: \\"ID!\\",
  };
}

class Query$getBook {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"getBook\\";
  static inputTypes = {
    id: \\"ID!\\",
  };
}

class Query$listBooks {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"listBooks\\";
  static inputTypes = {
    cursor: \\"Cursor\\",
    filter: \\"BookFilter\\",
  };
}

class Mutation$addReview {
  private static _: any;
  static typeProp = \\"mutation\\";
  static type = \\"Mutation\\";
  static prop = \\"addReview\\";
  static inputTypes = {
    input: \\"ReviewInput!\\",
  };
}

"
`;
